components/Buttons/src/MDCFloatingButton.m

```diff
@@ -60,79 +75,369 @@ - (instancetype)initWithFrame:(CGRect)frame shape:(MDCFloatingButtonShape)shape
+  self.shapeToHitAreaInsets[@(MDCFloatingButtonShapeDefault)]
+      = [NSValue valueWithUIEdgeInsets:UIEdgeInsetsZero];
+  self.shapeToHitAreaInsets[@(MDCFloatingButtonShapeMini)]
+      = [NSValue valueWithUIEdgeInsets:UIEdgeInsetsMake(-4, -4, -4, -4)];
+  super.contentEdgeInsets = [self defaultContentEdgeInsets];
+  super.hitAreaInsets = [self defaultHitAreaInsets];
+  _imageTitlePadding = 8;
+}
+
 #pragma mark - NSCoding
 
 - (instancetype)initWithCoder:(NSCoder *)aDecoder {
   self = [super initWithCoder:aDecoder];
   if (self) {
+    [self commonMDCFloatingButtonInit];
     _shape = [aDecoder decodeIntegerForKey:MDCFloatingButtonShapeKey];
+    if ([aDecoder containsValueForKey:MDCFloatingButtonMinimumSizeDictionaryKey]) {
+      _shapeToMinimumSize = [aDecoder decodeObjectForKey:MDCFloatingButtonMinimumSizeDictionaryKey];
+    }
+    if ([aDecoder containsValueForKey:MDCFloatingButtonMaximumSizeDictionaryKey]) {
+      _shapeToMaximumSize = [aDecoder decodeObjectForKey:MDCFloatingButtonMaximumSizeDictionaryKey];
+    }
+    if ([aDecoder containsValueForKey:MDCFloatingButtonContentEdgeInsetsDictionaryKey]) {
+      _shapeToContentEdgeInsets
+          = [aDecoder decodeObjectForKey:MDCFloatingButtonContentEdgeInsetsDictionaryKey];
+    }
+    if ([aDecoder containsValueForKey:MDCFloatingButtonHitAreaInsetsDictionaryKey]) {
+      _shapeToHitAreaInsets
+          = [aDecoder decodeObjectForKey:MDCFloatingButtonHitAreaInsetsDictionaryKey];
+    }
   }
   return self;
 }
 
 - (void)encodeWithCoder:(NSCoder *)aCoder {
   [super encodeWithCoder:aCoder];
-  [aCoder encodeInteger:_shape forKey:MDCFloatingButtonShapeKey];
+  [aCoder encodeInteger:self.shape forKey:MDCFloatingButtonShapeKey];
+  [aCoder encodeObject:self.shapeToMinimumSize forKey:MDCFloatingButtonMinimumSizeDictionaryKey];
+  [aCoder encodeObject:self.shapeToMaximumSize forKey:MDCFloatingButtonMaximumSizeDictionaryKey];
+  [aCoder encodeObject:self.shapeToContentEdgeInsets
+                forKey:MDCFloatingButtonContentEdgeInsetsDictionaryKey];
+  [aCoder encodeObject:self.shapeToHitAreaInsets
+                forKey:MDCFloatingButtonHitAreaInsetsDictionaryKey];
 }
 
 #pragma mark - UIView
 
-- (CGSize)sizeThatFits:(__unused CGSize)size {
+- (CGSize)intrinsicContentSize {
+  NSLog(@"intrinsicContentSize");
   switch (_shape) {
     case MDCFloatingButtonShapeDefault:
       return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
     case MDCFloatingButtonShapeMini:
       return CGSizeMake([[self class] miniDimension], [[self class] miniDimension]);
     case MDCFloatingButtonShapeLargeIcon:
       return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
+    case MDCFloatingButtonShapeExtendedTrailingIcon:
+    case MDCFloatingButtonShapeExtendedLeadingIcon: {
+      const CGSize intrinsicTitleSize
+          = [self.titleLabel sizeThatFits:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX)];
```

Comment:
Calling `sizeThatFits:` triggers an entire layout pass of the button due to the shadowLayer laying out its sublayers (apparently).